cmd/ingester/app/consumer/consumer.go

```diff
@@ -99,7 +103,18 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	msgMetrics := c.newMsgMetrics(pc.Partition())
 	var msgProcessor processor.SpanProcessor
 
-	for msg := range pc.Messages() {
+	// "Cast" channel from chan *sarama.ConsumerMessage to chan interface{}.
+	msgChannel := make(chan interface{})
+	go func() {
+		defer close(msgChannel)
+		for msg := range pc.Messages() {
+			msgChannel <- msg
+		}
+	}()
+
+	rateLimiter := newRateLimiter(msgChannel, c.maxReadsPerSecond)
+	for v := range rateLimiter.C {
```

Comment:
In theory, wouldn't this perform worse than the previous variant which called `ratelimiter.await` in this loop?

Reply:
In theory, yes. This is somewhat hard to determine, but I can attempt a benchmark of old vs. new version in the case where no rate limiting is applied (compares await with immediate return against wrapping inner channel). Otherwise, we will just be benchmarking the rate limit, which is useless.